File: ocaml/space-age/space_age.ml
open Base

type planet =
  | Mercury
  | Venus
  | Earth
  | Mars
  | Jupiter
  | Saturn
  | Neptune
  | Uranus
;;

let period planet years = match planet with
  | Mercury -> years /. 0.2408467
  | Venus -> years /. 0.61519726
  | Earth -> years
  | Mars -> years /. 1.8808158
  | Jupiter -> years /. 11.862615
  | Saturn -> years /. 29.447498
  | Uranus -> years /. 84.016846
  | Neptune -> years /. 164.79132
;;

let age_on planet age =
  let seconds = Float.of_int age in
  period planet (seconds /. 31557600.0)

File: haskell/space-age/src/SpaceAge.hs
module SpaceAge (Planet(..), ageOn) where

data Planet = Mercury
            | Venus
            | Earth
            | Mars
            | Jupiter
            | Saturn
            | Uranus
            | Neptune

period :: Planet -> Double
period Mercury = 0.2408467
period Venus = 0.61519726
period Earth = 1
period Mars = 1.8808158
period Jupiter = 11.862615
period Saturn = 29.447498
period Uranus = 84.016846
period Neptune = 164.79132

ageOn :: Planet -> Double -> Double
ageOn planet seconds = (seconds / 31557600) / period planet
